File: architecture/modules/storage/node/src/repository.ts
import * as fs from 'node:fs/promises';
import path from 'node:path';
import { StorageError,pageLimit,validKey,validSize,validContentType } from './contracts.ts';
import type {Page} from './contracts.ts';
import type {FileInfo} from './client.ts';
import { privateDirectory,readPrivate,atomicPrivate,withFileLock } from './private-files.ts';

export type FileState='pending'|'uploading'|'completing'|'ready'|'deleting'|'deleted'|'cancelled';
export interface FileRecord extends FileInfo {schemaVersion:1;state:FileState;ownerId:string;storeId:string;objectKey:string;tempKey:string;expiresAt:string;version:number;leaseUntil?:number}
export interface FileRepository {
  create(record:FileRecord):Promise<void>;
  get(id:string):Promise<FileRecord|undefined>;
  compareAndSwap(id:string,version:number,record:FileRecord):Promise<boolean>;
  list(ownerId:string,input?:{cursor?:string;limit?:number}):Promise<Page<FileRecord>>;
}
export function recordId(id:string):string {if(typeof id!=='string'||!/^[-a-f0-9]{36}$/.test(id))throw new StorageError('INVALID_INPUT','Invalid file ID');return id;}
export function parseRecord(value:string):FileRecord{
  const r=JSON.parse(value) as FileRecord;
  if(!r||typeof r!=='object'||r.schemaVersion!==1||!Number.isSafeInteger(r.version)||r.version<1||typeof r.ownerId!=='string'||!['pending','uploading','completing','ready','deleting','deleted','cancelled'].includes(r.state))throw new StorageError('CONFIGURATION','Invalid file metadata');
  recordId(r.id);validKey(r.objectKey);validKey(r.tempKey);validSize(r.size);validContentType(r.contentType);
  if(!r.ownerId||r.ownerId.length>255||typeof r.storeId!=='string'||!r.storeId||typeof r.name!=='string'||r.name.length>255||[r.createdAt,r.updatedAt,r.expiresAt].some(v=>typeof v!=='string'||!Number.isFinite(Date.parse(v)))||(r.leaseUntil!==undefined&&!Number.isFinite(r.leaseUntil)))throw new StorageError('CONFIGURATION','Invalid file metadata');return r;
}
export async function createFileRepository(directory:string):Promise<FileRepository>{
  const root=await privateDirectory(directory),file=(id:string)=>path.join(root,recordId(id)+'.json');
  async function get(id:string):Promise<FileRecord|undefined>{try{return parseRecord(await readPrivate(file(id)));}catch(e){if((e as NodeJS.ErrnoException).code==='ENOENT')return undefined;throw e;}}
  return {
    get,
    async create(record){parseRecord(JSON.stringify(record));await withFileLock(file(record.id)+'.lock',async()=>{if(await get(record.id))throw new StorageError('CONFLICT');await atomicPrivate(file(record.id),JSON.stringify(record));});},
    async compareAndSwap(id,version,record){parseRecord(JSON.stringify(record));return withFileLock(file(id)+'.lock',async()=>{const current=await get(id);if(!current||current.version!==version)return false;if(record.id!==id||record.version!==version+1||record.ownerId!==current.ownerId||record.storeId!==current.storeId)throw new StorageError('INVALID_INPUT','Invalid record transition');await atomicPrivate(file(id),JSON.stringify(record));return true;});},
    async list(ownerId,input={}){const limit=pageLimit(input.limit),items:FileRecord[]=[];if(input.cursor)recordId(input.cursor);const dir=await fs.opendir(root);for await(const entry of dir){if(!/^[-a-f0-9]{36}\.json$/.test(entry.name))continue;const r=await get(entry.name.slice(0,-5));if(r&&r.ownerId===ownerId&&r.state==='ready'&&r.id>(input.cursor||'')){items.push(r);items.sort((a,b)=>a.id<b.id?-1:a.id>b.id?1:0);if(items.length>limit+1)items.pop();}}return {items:items.slice(0,limit),...(items.length>limit?{cursor:items[limit-1].id}:{})};}
  };
}

File: architecture/modules/storage/go/repository.go
package storage

import (
	"context"
	"io"
	"os"
	"regexp"
	"sort"
	"strings"
)

type FileInfo struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Size        int64  `json:"size"`
	ContentType string `json:"contentType"`
	State       string `json:"state"`
	CreatedAt   string `json:"createdAt"`
	UpdatedAt   string `json:"updatedAt"`
}
type FileRecord struct {
	FileInfo
	SchemaVersion int    `json:"schemaVersion"`
	OwnerID       string `json:"ownerId"`
	StoreID       string `json:"storeId"`
	ObjectKey     string `json:"objectKey"`
	TempKey       string `json:"tempKey"`
	ExpiresAt     string `json:"expiresAt"`
	Version       int    `json:"version"`
	LeaseUntil    int64  `json:"leaseUntil,omitempty"`
}
type FilePage struct {
	Items  []FileInfo `json:"items"`
	Cursor string     `json:"cursor,omitempty"`
}
type RecordPage struct {
	Items  []FileRecord
	Cursor string
}
type Repository interface {
	Create(context.Context, FileRecord) error
	Get(context.Context, string) (FileRecord, error)
	CompareAndSwap(context.Context, string, int, FileRecord) (bool, error)
	List(context.Context, string, string, int) (RecordPage, error)
}

var idPattern = regexp.MustCompile(`^[a-f0-9-]{36}$`)

func validID(id string) error {
	if !idPattern.MatchString(id) {
		return fail("INVALID_INPUT")
	}
	return nil
}

type FileRepository struct{ root *os.Root }

func NewFileRepository(directory string) (*FileRepository, error) {
	root, e := privateRoot(directory)
	if e != nil {
		return nil, e
	}
	return &FileRepository{root}, nil
}
func (r *FileRepository) Close() error { return r.root.Close() }
func (r *FileRepository) Get(ctx context.Context, id string) (FileRecord, error) {
	var v FileRecord
	if e := ctx.Err(); e != nil {
		return v, Normalize(e)
	}
	if e := validID(id); e != nil {
		return v, e
	}
	e := readJSON(r.root, id+".json", &v)
	if e == nil && (v.ID != id || v.SchemaVersion != 1 || v.Version < 1) {
		e = fail("CONFIGURATION")
	}
	return v, e
}
func (r *FileRepository) Create(ctx context.Context, v FileRecord) error {
	if e := validID(v.ID); e != nil {
		return e
	}
	return fileLock(r.root, v.ID, func() error {
		_, e := r.Get(ctx, v.ID)
		if e == nil {
			return fail("CONFLICT")
		}
		if !IsCode(e, "NOT_FOUND") {
			return e
		}
		return atomicJSON(r.root, v.ID+".json", v)
	})
}
func (r *FileRepository) CompareAndSwap(ctx context.Context, id string, version int, v FileRecord) (bool, error) {
	if e := validID(id); e != nil {
		return false, e
	}
	ok := false
	e := fileLock(r.root, id, func() error {
		old, e := r.Get(ctx, id)
		if e != nil {
			return e
		}
		if old.Version != version {
			return nil
		}
		if v.ID != id || v.Version != version+1 || v.OwnerID != old.OwnerID || v.StoreID != old.StoreID {
			return fail("INVALID_INPUT")
		}
		if e = atomicJSON(r.root, id+".json", v); e != nil {
			return e
		}
		ok = true
		return nil
	})
	return ok, e
}
func (r *FileRepository) List(ctx context.Context, owner, cursor string, pageSize int) (RecordPage, error) {
	n, e := limit(pageSize)
	if e != nil {
		return RecordPage{}, e
	}
	if cursor != "" {
		if e = validID(cursor); e != nil {
			return RecordPage{}, e
		}
	}
	f, e := r.root.Open(".")
	if e != nil {
		return RecordPage{}, Normalize(e)
	}
	defer f.Close()
	items := []FileRecord{}
	for {
		entries, readErr := f.ReadDir(32)
		for _, entry := range entries {
			if !strings.HasSuffix(entry.Name(), ".json") {
				continue
			}
			id := strings.TrimSuffix(entry.Name(), ".json")
			if !idPattern.MatchString(id) {
				continue
			}
			v, e := r.Get(ctx, id)
			if e != nil {
				return RecordPage{}, e
			}
			if v.OwnerID == owner && v.State == "ready" && v.ID > cursor {
				items = append(items, v)
				sort.Slice(items, func(i, j int) bool { return items[i].ID < items[j].ID })
				if len(items) > n+1 {
					items = items[:n+1]
				}
			}
		}
		if readErr == io.EOF {
			break
		}
		if readErr != nil {
			return RecordPage{}, Normalize(readErr)
		}
	}
	result := RecordPage{Items: items}
	if len(items) > n {
		result.Cursor = items[n-1].ID
		result.Items = items[:n]
	}
	return result, nil
}
